### PR TITLE
Cover union types, type aliases, type guards

### LINKED ISSUE #37: type aliases
Possibly lumped with interfaces.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #37 and partially addresses #35

### CODE REVIEW

File: pages/Advanced Types.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
+# Introduction
+
+TODO
```
Comment (weswigham) [inline pages/Advanced Types.md]: Let me write one for you if that's okay:

> When thinking about types, sometimes it is useful to think about operations on those types -- if we have something which is both a `Fridge` and a `Toaster`, how can we best describe this amalgamation?
> Well, usually in the same way we just did -- as a `Fridge & Toaster`.
> This kind of basic type arithmetic can be incredibly useful in being more expressive with your types, helping you say what you mean without jumping through classical inheritance-based hoops.
> In TypeScript, we support three main productions which operate on types - [Unions](# Union Types), [Intersections](# Intersection Types), and [Aliases](# Type Aliases).
> Supporting these we have [Type Guards](# Type Guards), boolean-returning functions meant to indicate to the typesystem weather an argument is to be considered of a narrowed type in a given branch of code.
> Using these constructs correctly can enhance the semantic meaning of your types and help create a strongly typed, meaningful interface for your TypeScript program.

File: pages/Advanced Types.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+# Type Aliases
+
+Type aliases create a new name for a type.
+Type aliases are sometimes similar to interfaces, but can name primitives, unions, tuples, and any other types that you'd otherwise have to write by hand.
+
+```TypeScript
+type XCoord = number;
+type YCoord = number;
+
+type XCoord = { x: XCoord };
```
Comment (weswigham) [inline pages/Advanced Types.md]: Duplicate identifier 'XCoord'.

File: pages/Advanced Types.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+let pet = getSmallPet();
+pet.layEggs(); // okay
+pet.swim();    // errors
+```
+
+Union types can be a bit tricky here, but it just takes a bit of intuition to get used to.
+If a value has the type `A | B`, we only know for *certain* that it has members that both `A` *and* `B` have.
+On the other hand, in the example above `Bird` has a member named `fly`.
+We can't be sure whether a variable typed as `Bird | Fish` has a `fly` method.
+If the variable is really a `B` at runtime, then calling `pet.fly()` will fail.
```
Comment (sandersn) [inline pages/Advanced Types.md]: If it's really a `Fish`.

File: pages/Advanced Types.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+let pet = getSmallPet();
+pet.layEggs(); // okay
+pet.swim();    // errors
+```
+
+Union types can be a bit tricky here, but it just takes a bit of intuition to get used to.
+If a value has the type `A | B`, we only know for *certain* that it has members that both `A` *and* `B` have.
+On the other hand, in the example above `Bird` has a member named `fly`.
```
Comment (sandersn) [inline pages/Advanced Types.md]: I would reword as "In this example, `Bird` ..."

File: pages/Advanced Types.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    (<Bird>pet).fly();
+}
+```
+
+## User Defined Type Guards
+
+Notice that we had to use type assertions several times.
+It would be much better if once we performed the check, we could know the type of `pet` within each branch.
+
+It just so happens that TypeScript has something called a *type guard*.
```
Comment (sandersn) [inline pages/Advanced Types.md]: I would say something more like "To support this, Typescript has *type guards*.

File: pages/Advanced Types.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+A type guard is some expression that performs a runtime check that guarantees the type in some scope.
```
Comment (sandersn) [inline pages/Advanced Types.md]: "**an** expression"

File: pages/Advanced Types.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+We can write a type guard using a regular function:
```
Comment (sandersn) [inline pages/Advanced Types.md]: I would just say "a function".

File: pages/Advanced Types.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    return (<Fish>pet).swim !== undefined;
+}
+```
+
+Any time `isFish` is used on a variable name, TypeScript will know that variable has that specific type. 
+
+```TypeScript
+// Both calls to 'swim' and 'fly' are now okay.
+
+if (isFish(pet) {
```
Comment (sandersn) [inline pages/Advanced Types.md]: missing close parenthesis: `if (isFish(pet))

File: pages/Advanced Types.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+TODO
+
+## `instanceof` type guards
+
+TODO
+
+# Type Aliases
+
+Type aliases create a new name for a type.
+Type aliases are sometimes similar to interfaces, but can name primitives, unions, tuples, and any other types that you'd otherwise have to write by hand.
```
Comment (sandersn) [inline pages/Advanced Types.md]: A *little* more detail would be nice here -- my first thought was, "Why not get rid of interfaces then?" Specifically, I'd like a short explanation of what aliases can't do that interfaces can.

File: pages/Advanced Types.md [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+type Coordinate = XCoord | XYCoord | XYZCoord;
+type CoordList = Coordinate[];
+
+let coord: CoordList = [{ x: 10, y: 10}, { x: 0, y: 42, z: 10 }, { x: 5 }];
+```
+
+Aliasing doesn't actually create a new type - it creates a new *name* to refer to that type.
+So `10` is a perfectly valid `XCoord` and `YCoord` because they both just refer to `number`.
+Aliasing a primitive is not terribly useful, though it can be used as a form of documentation.
```
Comment (sandersn) [inline pages/Advanced Types.md]: Specifically, it sounds like aliases don't do the Haskell `newtype` thing where you can't assign between instances of the alias types:

```ts
var x: XCoord = 10;
var y: YCoord = x; // probably NO error
var z: XCoord = y; // probably NO error
```

It might be worthwhile to note that.

Comment (weswigham) [inline pages/Advanced Types.md]: I think that's what `*name*` was for.

File: pages/Advanced Types.md (hunk 1 above)
Comment (DanielRosenwasser) [inline pages/Advanced Types.md]: That's pretty great, though, I'm not 100% certain we're going to have intersection types here; it would be appropriate though.

**Comment by DanielRosenwasser** [comment]
@sandersn and @weswigham if you have feedback about recent changes, I'd be open to still making amendments.

**Comment by sandersn** [comment]
I still don't like "It just so happens Typescript has type guards". :persevere: 
That's the only thing -- the more recent changes look good.
